src/MultiFormats/MultiAddress.cs

```diff
@@ -0,0 +1,515 @@
+        /// </param>
+        /// <remarks>
+        ///   Reads the binary representation of <see cref="MultiAddress"/> from the <paramref name="buffer"/>.
+        ///   <para>
+        ///   The binary representation is a sequence of <see cref="NetworkProtocol">network protocols</see>.
+        ///   </para>
+        /// </remarks>
+        public MultiAddress(byte[] buffer)
+            : this()
+        {
+            if (buffer == null || buffer.Length == 0)
+                return;
+
+            Read(new MemoryStream(buffer, false));
+        }
+
+        /// <summary>
+        ///   Gets the peer ID of the multiaddress.
+        /// </summary>
+        /// <value>
+        ///   The <see cref="Ipfs.Core.Peer.Id"/> as a <see cref="MultiHash"/>.
+        /// </value>
+        /// <exception cref="Exception">
+        ///   When the last <see cref="Protocols">protocol</see>
+        ///   is not "ipfs" nor "p2p".
+        /// </exception>
+        /// <remarks>
+        ///   The peer ID is contained in the last protocol that
+        ///   is "ipfs" or "p2p".  For example, <c>/ip4/10.1.10.10/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC</c>.
+        /// </remarks>
+        public MultiHash PeerId
+        {
+            get
+            {
+                var protocol = Protocols
+                   .LastOrDefault(p => p.Name == "ipfs" || p.Name == "p2p");
+                if (protocol == null)
+                    throw new Exception($"'{this}' is missing the peer ID. Add the 'ipfs' or 'p2p' protocol.");
+
+                return protocol.Value;
+            }
+        }
+
+        /// <summary>
+        ///   Determines if the peer ID is present.
+        /// </summary>
+        /// <value>
+        ///   <b>true</b> if the peer ID present; otherwise, <b>false</b>.
+        /// </value>
+        /// <remarks>
+        ///   The peer ID is contained in the last protocol that
+        ///   is "ipfs" or "p2p".  For example, <c>/ip4/10.1.10.10/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC</c>.
+        /// </remarks>
+        public bool HasPeerId { get { return Protocols.Any(p => p.Name == "ipfs" || p.Name == "p2p"); } }
+
+        /// <summary>
+        ///   Gets a multiaddress that ends with the peer ID.
+        /// </summary>
+        /// <param name="peerId">
+        ///   The peer ID to end the multiaddress with.
+        /// </param>
+        /// <returns>
+        ///   Either the <c>this</c> multiadddress when it contains the
+        ///   <paramref name="peerId"/> or a new <see cref="MultiAddress"/>
+        ///   ending the <paramref name="peerId"/>.
+        /// </returns>
+        /// <exception cref="Exception">
+        ///   When the mulltiaddress has the wrong peer ID.
+        /// </exception>
+        public MultiAddress WithPeerId(MultiHash peerId)
+        {
+            if (HasPeerId)
+            {
+                var id = PeerId;
+                if (id != peerId)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)